### PR TITLE
#1535 generating test artifacts with a specified size for Maven

### LINKED ISSUE #1535: Implement support for generating test artifacts with a specified size for Maven
# Task Description

We need to be able to create test artifacts with a specified size.

# Classes of Interest

* `org.carlspring.strongbox.artifact.generator.MavenArtifactGenerator`
* `org.carlspring.strongbox.testing.artifact.MavenTestArtifact`

# Tasks

The following tasks will need to be carried out:
* [ ] Apply the necessary changes to the `org.carlspring.strongbox.artifact.generator.MavenArtifactGenerator`.
* [ ] Apply the necessary changes to the `org.carlspring.strongbox.testing.artifact.MavenTestArtifact`.
* [ ] Add tests for this.

# Task Relationships

This task:
* Is a sub-task of: #1534 
* Relates to: #1536 , #1537 , #1538 , #1540

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @fuss86
  * @sbespalov
  * @steve-todorov

### PR DESCRIPTION
**Pull request by wpater**
# Pull Request Description

Annotation used in tests improved to jar generation with given size.
This pull request closes #1535 

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [x] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [x] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

### CODE REVIEW

File: strongbox-storage/strongbox-storage-api/src/test/java/org/carlspring/strongbox/testing/artifact/TestArtifact.java [hunk 1]
```diff
@@ -66,7 +66,7 @@
     /**
      * {@link ArtifactGeneratorStrategy} class to use.
      */
-    Class<? extends ArtifactGeneratorStrategy<?>> strategy() default DefaultArtrifactGeneratorStrategy.class;    
+    Class<? extends ArtifactGeneratorStrategy<?>> strategy() default DefaultArtifactGeneratorStrategy.class;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/test/java/org/carlspring/strongbox/testing/artifact/TestArtifact.java]: :+1: Thanks for spotting this typo in the class' name! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/MavenArtifactGenerator.java [hunk 2]
```diff
@@ -264,13 +282,14 @@ private void createMavenPropertiesFile(Artifact artifact, ZipOutputStream zos)
         zos.closeEntry();
     }
 
-    private void createRandomSizeFile(ZipOutputStream zos)
+    private void createFile(ZipOutputStream zos,
+                            long size)
             throws IOException
     {
-        ZipEntry ze = new ZipEntry("random-size-file");
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/MavenArtifactGenerator.java]: I think you should add an if check here... If a size hasn't been specified, create the random-sized file (with the limit as it was previously).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java [hunk 3]
```diff
@@ -65,4 +66,9 @@
      */
     String packaging() default "jar";
 
+    /**
+     * Additional artifact size in bytes.
+     */
+    @AliasFor(annotation = TestArtifact.class)
+    long size() default 1024;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java]: I think this should default to `0` and there should be a check for this in the generator. (`1024` is an most-likely not enough in order to create a proper jar with a manifest and all that).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/MavenArtifactGenerator.java (hunk 2 above)
Comment (wpater) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/MavenArtifactGenerator.java]: In all cases where size is not specified it is randomed - for example in `public void generate(Artifact artifact)`. I will change that random limit to previous value (1000000).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java (hunk 3 above)
Comment (wpater) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java]: This size is size of additional file in artifact - it's not size of whole artifact. After adding files and pom to archive checksum will be generated and size of whole jar will be bigger than given size. It's more like minimal size of artifact and at the end it should be bigger than given size.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/MavenArtifactGenerator.java [hunk 4]
```diff
@@ -333,14 +352,14 @@ private void generateChecksumsForArtifact(File artifactFile)
             Path checksumPath = artifactPath.resolveSibling(artifactPath.getFileName() + EncryptionAlgorithmsEnum.MD5.getExtension());
             try (OutputStream os = newOutputStream(checksumPath.toFile()))
             {
-                IOUtils.write(md5, os, Charset.forName("UTF-8"));
+                IOUtils.write(md5, os, StandardCharsets.UTF_8);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/MavenArtifactGenerator.java]: :+1: Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java (hunk 3 above)
Comment (carlspring) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java]: Yes, thanks for reminding me of this. The size is the one of the random binary file that's included inside the `jar`.

**Comment by carlspring** [comment]
Hi @wpater !

Thank you for your contribution!

Would you mind signing the ICLA, as described in the [Contributing](https://github.com/strongbox/strongbox/blob/master/CONTRIBUTING.md) page?

Kind regards,

Martin

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java (hunk 3 above)
Comment (wpater) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java]: How much then? 1000000?

Comment (carlspring) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/MavenTestArtifact.java]: Yeah, I think that'll be better.

**Comment by carlspring** [comment]
Thanks for signing the ICLA and for your contribution! Congratulations on your first pull request! :)

**Pull request merged**